source/sliding_puzzle.py

```diff
@@ -124,7 +139,67 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        random.shuffle(self.rows)
+        # empty one of the rows
+        random.choice(random.choice(self.rows)).clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece._data == []
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
+
+    def move_up(self) -> None:
+        """Move the piece above the empty piece down"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the first row
+        top_row = self.rows[0]
+        if empty_pos.index in [piece.index for piece in top_row]:
+            return
+        self._swap_pieces(x1=empty_pos.x, y1=empty_pos.y, x2=empty_pos.x - 1, y2=empty_pos.y)
```

Comment:
```suggestion
        self._swap_pieces(empty_pos.x, empty_pos.y, empty_pos.x - 1, empty_pos.y)
```

Reply:
I still think we should just pass it as arguments, it's much less to read

Reply:
for the sake of clarity, I think the keyword should stay, it may be easier to read without them, but they are still a usefull piece of info

Reply:
Not really; if it's harder to read with them then that's not clarity. I don't see what additional info they provide that we don't know. We know they take 4 coordinates and swaps the pairs